### PR TITLE
AMP-28731: Expose ids in API

### PR DESCRIPTION
**Pull request by developster**

### CODE REVIEW

File: amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/FieldAccessor.java [hunk 1]
```diff
@@ -3,7 +3,9 @@
 /**
  * @author Octavian Ciubotaru
  */
-public interface FieldValueReader {
+public interface FieldAccessor {
 
     Object get(Object targetObject);
+
+    void set(Object targetObject, Object value);
```
[the remark this answers is not in the document]
Comment (developster) [reply amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/FieldAccessor.java]: No. But DiscriminatedFieldAccessor is a good place to set the discrimination field. That way ObjectImporter will be less concerned about discrimination.

File: amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/ObjectImporter.java [hunk 2]
```diff
@@ -410,42 +373,162 @@ private boolean validateSubElements(APIField fieldDef, Object newParent, Object
...
+
+                    // match elements by ids and import them
+                    Set<Object> jsonIds = new HashSet<>();
+                    for (Map<String, Object> newChild : childrenNewValues) {
+                        Object jsonId = newChild.get(idField.getFieldName());
+                        jsonId = convert(idField.getApiType().getType(), jsonId);
+
+                        if (jsonId != null) {
+                            jsonIds.add(jsonId);
+                        }
```
[the remark this answers is not in the document]
Comment (developster) [reply amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/ObjectImporter.java]: Ok.

File: amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/ObjectImporter.java [hunk 3]
```diff
@@ -410,42 +373,162 @@ private boolean validateSubElements(APIField fieldDef, Object newParent, Object
...
+                    }
+
+                    removeByIdExcept(idField, newFieldValueCollection, jsonIds);
+                } else if (fieldDef.isDiscriminatedObject()) {
+                    Collection newFieldValueCollection = (Collection) newFieldValue;
+                    Map<String, Object> newChild = childrenNewValues.isEmpty()
+                            ? Collections.emptyMap() : childrenNewValues.get(0);
+                    if (newFieldValueCollection.size() > 1) {
+                        throw new RuntimeException("Expected one element at most! At: " + fieldPath);
+                    }
```
[the remark this answers is not in the document]
Comment (developster) [reply amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/ObjectImporter.java]: This branch is invoked only when json field is one object but internally it is discriminated and part of a collection. DiscriminatedFieldReader in such case must return a collection with just one item. This exception will be thrown when for example an activity has two PPC amounts. Added this to catch either configuration or data errors.

File: amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/FieldAccessor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (developster) [reply amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/FieldAccessor.java]: I think it is best to analyze this change once we refactor ObjectImporter to use APIField and FieldAccessor instead of direct calls to reflection.

File: amp/src/test/java/org/digijava/kernel/ampapi/endpoints/activity/ObjectImporterTest.java [hunk 4]
```diff
@@ -87,4 +396,233 @@ public void testBackReferences() throws IOException {
...
+
+        Parent parent = new Parent();
+        parent.addChild(new Child(1L, "A", "A"));
+        parent.addChild(new Child(1L, "A", "A"));
+
+        importer.validateAndImport(parent, json);
+    }
+
+    @Test
+    public void testImportableFieldsAreClearedIfMissingInJson() {
```
[the remark this answers is not in the document]
Comment (developster) [reply amp/src/test/java/org/digijava/kernel/ampapi/endpoints/activity/ObjectImporterTest.java]: Added on purpose, test will have to be adjusted.

File: amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/ObjectImporter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (nmandrescu) [reply amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/ObjectImporter.java]: Makes sense if the list comes from `DiscriminatedFieldReader`. It was hard to follow, so I had a feeling this collection was about the entire list, that can have both PPC and RPC in some countries.
